### ISSUE TITLE
Works with CPU but not CUDA

### ISSUE THREAD

**Issue by MustangHX** [ultralytics/yolov5#1391]
## 🐛 Bug
There are detections with CPU only, but not with GPU.  Also, no labels produced with GPU. This is on a windows laptop with conda python environmnet.

## To Reproduce (REQUIRED)
 git clone https://github.com/ultralytics/yolov5.git
pip install -r requirements.txt
 python .\detect.py --source .\f35.mp4 --save-txt

Output:
 python .\detect.py --source .\f35.mp4 --save-txt
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=True, source='.\\f35.mp4', update=False, view_img=False, weights='yolov5s.pt')
Using torch 1.7.0 CUDA:0 (GeForce GTX 1660 Ti, 6144MB)

Fusing layers...
Model Summary: 140 layers, 7.45958e+06 parameters, 0 gradients
video 1/1 (1/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.031s)
video 1/1 (2/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.029s)
video 1/1 (3/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.029s)
video 1/1 (4/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.029s)
video 1/1 (5/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.024s)
video 1/1 (6/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.022s)
video 1/1 (7/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.022s)
video 1/1 (8/294) C:\Users\feynm\yolov5\f35.mp4: 384x640 Done. (0.023s)
```


## Expected behavior
There should be detections, similar to the result when running on CPU.

## Environment
If applicable, add screenshots to help explain your problem.

 - OS: [windows 10]
 - GPU [1660 Ti]

**Comment by glenn-jocher** [comment]
@MustangHX we have a baseline video for bug reproduction. Are you able to reproduce locally and in Colab with this video?
https://github.com/ultralytics/yolov5/releases/download/v1.0/decelera_landscape.mov

Please ensure you meet all dependency requirements if you are attempting to run YOLOv5 locally.  If in doubt, create a new virtual Python 3.8 environment, clone the latest repo (**code changes daily**), and `pip install -r requirements.txt` again. We also highly recommend using one of our verified environments below. 


## Requirements

**Python 3.8** or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including **`torch>=1.6`**. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by glenn-jocher** [comment]
GPU: 
<img width="1384" alt="Screenshot 2020-11-13 at 19 15 49" src="https://user-images.githubusercontent.com/26833433/99106550-aff87900-25e4-11eb-81af-86465d349d5b.png">

**Comment by glenn-jocher** [comment]
CPU:

<img width="1384" alt="Screenshot 2020-11-13 at 19 16 23" src="https://user-images.githubusercontent.com/26833433/99106578-bdadfe80-25e4-11eb-8f78-e55091bdda30.png">

**Comment by MustangHX** [comment]
The baseline video gives the same result, no detections.

Interestingly, I found that I can detect objects from my webcam stream normally, using either GPU or CPU.

**Comment by glenn-jocher** [comment]
@MustangHX understood. I will paste you our standard bug response below. The fundamental requirement for us to investigate is reproducibility. We need a reproducible example in one of the common environments below to start.

Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and then install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the **Bug Report template**, providing screenshots and **minimum viable code to reproduce your issue**. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.6`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by billrune** [comment]
i got same problem.  work fine with CPU but not with the GPU.  no object detected out with GPU. i installed with the latest version display driver 461.40. and the cuda_11.1.0_456.43_win10.exe. 

C:\Users\Administrator\yolov5_new>python detect.py --device="cpu" --weights="weights/yolov5s.pt"
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='cpu', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=False, source='data/images', update=False, view_img=False, weights=['weights/yolov5s.pt'])
YOLOv5 v4.0-100-gca5b10b torch 1.7.1+cu110 CPU

Fusing layers...
Model Summary: 224 layers, 7266973 parameters, 0 gradients, 17.0 GFLOPS
image 1/2 C:\Users\Administrator\yolov5_new\data\images\bus.jpg: 640x480 4 persons, 1 bus, Done. (0.174s)
image 2/2 C:\Users\Administrator\yolov5_new\data\images\zidane.jpg: 384x640 2 persons, 1 tie, Done. (0.136s)
Results saved to runs\detect\exp2
Done. (0.401s)

C:\Users\Administrator\yolov5_new>python detect.py --device="" --weights="weights/yolov5s.pt"
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=False, source='data/images', update=False, view_img=False, weights=['weights/yolov5s.pt'])
YOLOv5 v4.0-100-gca5b10b torch 1.7.1+cu110 CUDA:0 (GeForce GTX 1650 SUPER, 4096.0MB)

Fusing layers...
Model Summary: 224 layers, 7266973 parameters, 0 gradients, 17.0 GFLOPS
image 1/2 C:\Users\Administrator\yolov5_new\data\images\bus.jpg: 640x480 Done. (0.033s)
image 2/2 C:\Users\Administrator\yolov5_new\data\images\zidane.jpg: 384x640 Done. (0.031s)
Results saved to runs\detect\exp3
Done. (0.119s)

**Comment by glenn-jocher** [comment]
@billrune I think this was traced down to a windows driver problem in a different issue thread (possibly no solution). You may want to try a different environment for now:

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by billrune** [comment]
@glenn-jocher okay! thank you so much!

**Comment by FareedKhan-dev** [comment]
Source of the answer: [link](https://stackoverflow.com/questions/74696085/yolo-v7-not-detecting-objects-on-image)

Since when you are working with GPU, It allows half precision by default, which you can change by editing your `detect.py` file.

Go to `detect.py` file and not exactly sure but on line 31, you will see this line of code:

`half = device.type != 'cpu'  # half precision only supported on CUDA`

Replace that line with

`half = False`

and then save `default.py` file.

Now when you are using your detection command, make sure to use `--device 0` that indicates your GPU must be utilize for detection.

`python .\detect.py --source .\f35.mp4  --device 0 --save-txt`

**Comment by glenn-jocher** [comment]
@FareedKhan-dev thanks for sharing that suggestion. However, the default setting *`half = False`* is typically already in configuration files. This setting disables half-precision if your GPU doesn't support it.

Regarding the `--device 0` argument in the detection command, it is typically not needed, as YOLOv5 automatically utilizes the GPU if present. Explicitly specifying `--device 0` should not be necessary in most cases.

Please feel free to keep the discussions and suggestions coming! It's invaluable to have the YOLO community collaborating on these solutions.
